### DISCUSSION TITLE
How to use WeakReference in F#

### DISCUSSION THREAD

**Discussion by xp44mm** [fsharp/fslang-suggestions#1236] [Library]
This is a piece of C# code:

```C#
var obj = new object();
var weak = new WeakReference(obj);
GC.Collect();
Console.WriteLine("IsAlive: {0} obj!=null is {1}", weak.IsAlive,obj!=null);

obj = null;
GC.Collect();
Console.WriteLine("IsAlive: {0}", weak.IsAlive);
```

This is the output you'll see when running the example:

```
IsAlive: True obj!=null is True
IsAlive: False
```
This is the F# code I translated:

```fsharp
    let mutable obj = new obj()
    let weak = new WeakReference(obj)
    GC.Collect()
    Console.WriteLine("IsAlive: {0} obj!=null is {1}", weak.IsAlive, obj<>null)

    obj <- null
    GC.Collect()
    Console.WriteLine("IsAlive: {0}", weak.IsAlive)
```

The problem is that the reference is still alive after `obj` becomes `null`.

**Comment by vzarytovskii** [comment]
I would ask this question on stackoverflow, since not that many people read this repository's discussions.

**Comment by jl0pd** [comment]
This happens because of how F# compiles top-level statements. Instead of placing local variables into locals, it places them into static fields. Moving all code into function solves the issue

```fsharp
open System

let main() =
    let mutable obj = new obj()
    let weak = new WeakReference(obj)
    GC.Collect()
    Console.WriteLine("IsAlive: {0} obj!=null is {1}", weak.IsAlive, obj<>null)

    obj <- null
    GC.Collect()
    Console.WriteLine("IsAlive: {0}", weak.IsAlive)
    
main()
```

There's awesome tool for inspecting compilation result: https://sharplab.io/

**Reply by xp44mm** [reply]
The code is placed in a local function, and running it is still an unexpected result.

**Reply by Xyncgas** [reply]
GC.Collect() doesn't always collect, Maybe that's why
